sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala

```diff
@@ -66,24 +66,28 @@ case class OrcPartitionReaderFactory(
   override def buildReader(file: PartitionedFile): PartitionReader[InternalRow] = {
     val conf = broadcastedConf.value.value
 
-    val resultSchemaString = OrcUtils.orcTypeDescriptionString(resultSchema)
-    OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)
     OrcConf.IS_SCHEMA_EVOLUTION_CASE_SENSITIVE.setBoolean(conf, isCaseSensitive)
 
     val filePath = new Path(new URI(file.filePath))
 
     val fs = filePath.getFileSystem(conf)
     val readerOptions = OrcFile.readerOptions(conf).filesystem(fs)
-    val requestedColIdsOrEmptyFile =
+    val resultedColPruneInfo =
       Utils.tryWithResource(OrcFile.createReader(filePath, readerOptions)) { reader =>
         OrcUtils.requestedColumnIds(
           isCaseSensitive, dataSchema, readDataSchema, reader, conf)
       }
 
-    if (requestedColIdsOrEmptyFile.isEmpty) {
+    if (resultedColPruneInfo.isEmpty) {
       new EmptyPartitionReader[InternalRow]
     } else {
-      val requestedColIds = requestedColIdsOrEmptyFile.get
+      val (requestedColIds, canPruneCols) = resultedColPruneInfo.get
+      val resultSchemaString = if (canPruneCols) {
+        OrcUtils.orcTypeDescriptionString(resultSchema)
+      } else {
+        OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields))
+      }
+      OrcConf.MAPRED_INPUT_SCHEMA.setString(conf, resultSchemaString)
```

Comment:
To avoid duplicated code, we can move these to the `requestedColumnIds` method
```
def requestedColumnIds(..., partitionSchema: StructType): Option[Array[Int]] = {
  ...
  if (orcFieldNames.isEmpty) {
    None
  } else if (orcFieldNames.forall(_.startsWith("_col"))) {
    OrcConf.MAPRED_INPUT_SCHEMA.setString(conf,
      OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields)))
    ...
  } else {
    OrcConf.MAPRED_INPUT_SCHEMA.setString(conf,
      OrcUtils.orcTypeDescriptionString(StructType(requiredSchema.fields ++ partitionSchema.fields)))
    ...
  }
}
```

Reply:
@cloud-fan - In this we need to return the resultSchemaString from this method  Option[(Array[Int], String)]
 which is for else if (orcFieldNames.forall(_.startsWith("_col"))) {
 val resultSchemaString = OrcUtils.orcTypeDescriptionString(StructType(dataSchema.fields ++ partitionSchema.fields)

else
val resultSchemaString =  OrcUtils.orcTypeDescriptionString(StructType(requiredSchema.fields ++ partitionSchema.fields)))

since we are using this resultSchemaString in

batchReader.initBatch(
        TypeDescription.fromString(resultSchemaString),
        resultSchema.fields,

Reply:
adding a new helper method is also good.

Reply:
Added the helper method